NIFI-12452 Improve support for DescribedValue

EndzeitBegins opened the pull request:
<!-- Licensed to the Apache Software Foundation (ASF) under one or more -->
<!-- contributor license agreements.  See the NOTICE file distributed with -->
<!-- this work for additional information regarding copyright ownership. -->
<!-- The ASF licenses this file to You under the Apache License, Version 2.0 -->
<!-- (the "License"); you may not use this file except in compliance with -->
<!-- the License.  You may obtain a copy of the License at -->
<!--     http://www.apache.org/licenses/LICENSE-2.0 -->
<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Summary

Improves usability of `DescribedValue` for property declaration, during tests and property value interpretation.
Makes exemplary use of new / modified APIs for `ListenHTTP`.

[NIFI-12452](https://issues.apache.org/jira/browse/NIFI-12452)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 21

### Licensing

- [x] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [x] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] Documentation formatting appears as expected in rendered files

Lehel44 commented:
Thanks @EndzeitBegins for the improvement! I will review

EndzeitBegins commented on nifi-api/src/main/java/org/apache/nifi/components/PropertyDescriptor.java:
```diff
@@ -313,6 +313,27 @@ public Builder defaultValue(final String value) {
+         * @param describedValue default value holder
+         * @return the builder
+         */
+        public Builder defaultDescribedValue(final DescribedValue describedValue) {
```
Yes, this was my initial idea as well. However, this prohibits the user of `.defaultValue(null)` as the compiler cannot differentiate in that case. 
If we're okay with that, I'm more than happy to make it an overloaded method. What do you think?

markap14 replied:
I definitely think it's fine to require `.defaultValue((String) null)`. But I'm not sure this is ever needed, since `null` is always the default unless otherwise specified.

EndzeitBegins commented:
Thank you for your feedback @exceptionfactory and @markap14.

I renamed `defaultDescribedValue(DescribedValue)` to `defaultValue(DescribedValue)` making it an overload of the existing `defaultValue(String)`.

In this process I also removed all calls to `defaultValue` that statically provide `null` I could find in the codebase.  All these calls were effectively doing nothing, as the existing implementation of `defaultValue` just returns the `PropertyDescriptor.Builder` without any modification upon receiving `null`. 
I assumed this is a more straightforward change than keeping the calls but as `defaultValue((String) null)`.

exceptionfactory reviewed (commented):
Thanks for making the adjustments @EndzeitBegins.

After reviewing the changes required for renaming the `describeDefaultValue()` method, I have some concerns about introducing this as a breaking change.

Although the `defaultValue(null)` calls are not necessary, the fact of their existence highlights the possibility of other extension components doing something similar. Making this change would also require code changes when recompiling against a new version. This may be minor enough not to be a concern, but it would need to be highlighted.

Another alternative would be to have the existing method take an `Object` argument, and then use `instanceof` checks to determine whether it is a String, `DescribedValue`, something unsupported, or null.

Thoughts on approaches @markap14 and @EndzeitBegins?

markap14 commented:
My preference is to keep the name `defaultValue` and add the override. I would avoid allowing for a `defaultValue` that accepts an `Object`. While this could be a change that breaks backward compatibility for a few, I think it's a pretty niche concern and still fair game for a 2.0, as long as it's mentioned in the Migration Guide. Also of note, custom code could be updated to use `.defaultValue( (String) null )` even in 1.x so there would be no need to maintain two different versions of the code, etc. So I think it's fair game.

EndzeitBegins commented:
Sounds reasonable to me. Thank you both @exceptionfactory & @markap14 for your valuable feedback. From my understanding the current PR proposal with the overload of `defaultValue` corresponds to the desired approach then.

Could one of you take on the task to enhance the [migration guide](https://cwiki.apache.org/confluence/plugins/servlet/mobile?contentId=57905503#MigrationGuidance-Migratingto2.0.0-M1) then? As far as I'm aware sole contributors cannot propose or make changes to Confluence?

I've seen that there are conflicts with `main` now. I assume this is due to `HttpNotificationService` being removed by #8104. Should I rebase and force-push this PR to be compatible again or should I let the commiter resolve the issue?

exceptionfactory commented:
Thanks for the reply @EndzeitBegins. Yes, if you can rebase and force push to resolve the conflict with HttpNotificationService being removed, that would be helpful.

I can start a new section in the Migration Guide.

EndzeitBegins commented:
I just rebased & squashed the commits, @exceptionfactory.

exceptionfactory reviewed (approved):
Thanks again for your work on this improvement @EndzeitBegins. I added a new section to the [Migration Guidance](https://cwiki.apache.org/confluence/display/NIFI/Migration+Guidance#MigrationGuidance-Migratingto2.0.0) mentioning the changes. +1 merging

EndzeitBegins commented:
Thank you for taking on the merge process and adjusting the Migration Guidance, @exceptionfactory. Have a nice week.
